src/transport.rs

```diff
@@ -11,51 +11,323 @@
+    ) {
+        let topic_listener_map = Arc::new(RwLock::new(HashMap::new()));
+
+        let client = UPClientMqtt {
+            mqtt_client: Arc::new(MockMqttClient {}),
+            subscription_topic_map: Arc::new(RwLock::new(HashMap::new())),
+            topic_listener_map,
+            authority_name: "VIN.vehicles".to_string(),
+            client_type: UPClientMqttType::Device,
+            free_subscription_ids: Arc::new(RwLock::new((1..10).collect())),
+        };
+
+        let listener = Arc::new(SimpleListener {});
+
+        let source_uri = UUri::from_str(source_filter).expect("Expected a valid source value");
+
+        let sink_uri = sink_filter.map(|s| UUri::from_str(s).expect("Expected a valid sink value"));
+
+        let result = client
+            .register_listener(&source_uri, sink_uri.as_ref(), listener.clone())
+            .await;
+
+        if result.is_err() {
+            assert_eq!(result.err().unwrap(), expected_error.unwrap());
+        } else {
+            assert!(result.is_ok());
+        }
+
+        let topic_map = client.topic_listener_map.read().await;
+
+        assert!(topic_map.contains_key(expected_topic));
+
+        let listeners = topic_map.get(expected_topic).unwrap();
+
+        assert!(listeners.contains(&ComparableListener::new(listener)));
+    }
+
+    #[test_case("//VIN.vehicles/A8000/2/8A50", None, "d/VIN.vehicles/A8000/2/8A50", None; "Unregister listener success")]
+    #[test_case("//VIN.vehicles/A8000/2/8A50", Some("//VIN.vehicles/B8000/3/0"), "d/VIN.vehicles/A8000/2/8A50/VIN.vehicles/B8000/3/0", None; "Unregister listener with sink success")]
+    #[async_std::test]
+    async fn test_unregister_listener(
+        source_filter: &str,
+        sink_filter: Option<&str>,
+        expected_topic: &str,
+        expected_error: Option<UStatus>,
+    ) {
+        let topic_listener_map = Arc::new(RwLock::new(HashMap::new()));
+
+        let listener = Arc::new(SimpleListener {});
+        let comparable_listener = ComparableListener::new(listener.clone());
+
+        topic_listener_map.write().await.insert(
+            expected_topic.to_string(),
+            [comparable_listener.clone()].iter().cloned().collect(),
+        );
+
+        let client = UPClientMqtt {
+            mqtt_client: Arc::new(MockMqttClient {}),
+            subscription_topic_map: Arc::new(RwLock::new(HashMap::new())),
+            topic_listener_map,
+            authority_name: "VIN.vehicles".to_string(),
+            client_type: UPClientMqttType::Device,
+            free_subscription_ids: Arc::new(RwLock::new((1..10).collect())),
+        };
+
+        let source_uri = UUri::from_str(source_filter).expect("Expected a valid source value");
+
+        let sink_uri = sink_filter.map(|s| UUri::from_str(s).expect("Expected a valid sink value"));
+
+        let result = client
+            .unregister_listener(&source_uri, sink_uri.as_ref(), listener.clone())
+            .await;
+
+        if result.is_err() {
```

Comment:
What if accidentally a unit test is written s.t. result is not an error, but this was intended to be an error case?

Could we use whether expected_error is not None in order to account for that case?

Reply:
yes I think we can. Added `assert!(expected_error.is_none());` in the else condition.